### PR TITLE
Fix Update command to deny similar inputs and Add more Junit test for update command.

### LINKED ISSUE #117: [PE-D] No Update Notification
I think it is better to notify the user that updated date is the same as the one in the list, and no update made.

-------------
Labels: `type.FeatureFlaw` `severity.Low`
original: lihao-InfoSec/ped#5

### PR DESCRIPTION
**Pull request by limjierui**
Fixes #117 

Fix Update command to deny similar inputs.

Add more Junit test for update command.

### CODE REVIEW

File: src/main/java/seedu/mindmymoney/command/UpdateCommand.java [hunk 1]
```diff
@@ -121,7 +121,11 @@ public void updateExpenditure() throws MindMyMoneyException {
             }
             LocalDate date = LocalDate.parse(inputTime, DateTimeFormatter.ofPattern("dd/MM/yyyy"));
             String newTime = date.format(DateTimeFormatter.ofPattern("dd/MM/yyyy"));
-
+            if (isSimilarExpenditure(indexToUpdate, newPaymentMethod, newCategory, newDescription, newAmountAsFloat,
+                    newTime)) {
+                throw new MindMyMoneyException("Expense fields to be updated is similar to the expense in the list.\n"
+                        + "Please make sure the field descriptions you want to change are different.");
+            }
             //Create new expenditure object to substitute in
```
Comment (khseah) [inline src/main/java/seedu/mindmymoney/command/UpdateCommand.java:129]: If I recall correctly, comments must have a whitespace after //
```suggestion
            // Create new expenditure object to substitute in
```

File: src/main/java/seedu/mindmymoney/command/UpdateCommand.java [hunk 2]
```diff
@@ -140,6 +144,29 @@ public void updateExpenditure() throws MindMyMoneyException {
...
+     * @param newTime new time field to be updated.
+     * @return true if fields are similar, false otherwise.
+     */
+    public boolean isSimilarExpenditure(int index, String newPaymentMethod, String newCategory, String newDescription,
+                             float newAmountAsFloat, String newTime) {
+        if (expenditureList.get(index).getPaymentMethod().equals(newPaymentMethod)
+                && expenditureList.get(index).getCategory().equals(newCategory)
+                && expenditureList.get(index).getDescription().equals(newDescription)
+                && expenditureList.get(index).getAmount() == newAmountAsFloat
+                && expenditureList.get(index).getTime().equals(newTime)) {
```
Comment (khseah) [inline src/main/java/seedu/mindmymoney/command/UpdateCommand.java:164]: Maybe can refactor out each condition to improve SLAP and readability

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 3]
```diff
@@ -25,17 +31,19 @@ void updateCommand_updateExpenditure_listUpdated() {
...
         } catch (MindMyMoneyException e) {
             System.out.println(e.getMessage());
             fail();
         }
     }
 
-    /** Assert that an invalid command will throw an exception. */
+    /**
+     * Assert that an invalid update expenditure command will throw an exception.
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:43]: Refine javadoc format
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 4]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
         UpdateCommand updateCommand = new UpdateCommand(input, testUser);
         assertThrows(MindMyMoneyException.class, updateCommand::executeCommand);
     }
+
+    /**
+     * Assert that when the update command fields are similar to the expenditure in the list, an exception is thrown.
+     *
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:58]: Can delete this line
```suggestion
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 5]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:59]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 6]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        testUser.setExpenditureListArray(new ExpenditureList());
+        testUser.getExpenditureListArray().add(testExpenditure);
+        String input = "1 /pm cash /c food /d porridge /a 5 /t 01/04/2022";
+        UpdateCommand updateCommand = new UpdateCommand(input,testUser);
+        assertThrows(MindMyMoneyException.class, updateCommand::executeCommand);
+    }
+
+    /**
+     * Assert that the command can update the credit card list.
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:74]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 7]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        CreditCard testCreditCard = new CreditCard("DBS", 2, 1000, 1000);
+        CreditCard newCreditCard = new CreditCard("DBS", 5, 2000, 2000);
+        User testUser = new User();
+        testUser.setCreditCardListArray(new CreditCardList());
+        testUser.getCreditCardListArray().add(testCreditCard);
+        String input = "/cc 1 /n DBS /cb 5 /cl 2000 /bal 2000";
+        UpdateCommand updateCommand = new UpdateCommand(input, testUser);
+        try {
+            updateCommand.executeCommand();
+            assertEquals(testUser.getCreditCardListArray().get(0).toString(), newCreditCard.toString());
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:86]: Remove magic literal, remember to import the constant
```suggestion
            assertEquals(testUser.getCreditCardListArray().get(INDEX_OF_FIRST_ITEM).toString(), newCreditCard.toString());
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 8]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        } catch (MindMyMoneyException e) {
+            System.out.println(e.getMessage());
+            fail();
+        }
+    }
+
+    /**
+     * Assert that an invalid update credit card command will throw an exception.
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:95]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 9]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        testUser.getCreditCardListArray().add(testCreditCard);
+        String input = "/cc invalid input";
+        UpdateCommand updateCommand = new UpdateCommand(input, testUser);
+        assertThrows(MindMyMoneyException.class, updateCommand::executeCommand);
+    }
+
+    /**
+     * Assert that when the update command fields are similar to the credit card in the list, an exception is thrown.
+     *
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:110]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 10]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        testUser.setCreditCardListArray(new CreditCardList());
+        testUser.getCreditCardListArray().add(testCreditCard);
+        String input = "/cc 1 /n DBS /cb 2 /cl 1000 /bal 1000";
+        UpdateCommand updateCommand = new UpdateCommand(input,testUser);
+        assertThrows(MindMyMoneyException.class, updateCommand::executeCommand);
+    }
+
+    /**
+     * Assert that the command can update the income list.
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:124]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 11]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        Income testIncome = new Income(1000, "Salary");
+        Income newIncome = new Income(3000, "Salary");
+        User testUser = new User();
+        testUser.setIncomeListArray(new IncomeList());
+        testUser.getIncomeListArray().add(testIncome);
+        String input = "/i 1 /a 3000 /c salary";
+        UpdateCommand updateCommand = new UpdateCommand(input, testUser);
+        try {
+            updateCommand.executeCommand();
+            assertEquals(testUser.getIncomeListArray().get(0).toString(), newIncome.toString());
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:136]:
```suggestion
            assertEquals(testUser.getIncomeListArray().get(INDEX_OF_FIRST_ITEM).toString(), newIncome.toString());
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 12]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        } catch (MindMyMoneyException e) {
+            System.out.println(e.getMessage());
+            fail();
+        }
+    }
+
+    /**
+     * Assert that an invalid update income command will throw an exception.
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:145]:
```suggestion
     **/
```

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java [hunk 13]
```diff
@@ -44,4 +52,120 @@ void updateCommand_invalidInput_exceptionThrown() {
...
+        testUser.getIncomeListArray().add(testIncome);
+        String input = "/i invalid input";
+        UpdateCommand updateCommand = new UpdateCommand(input, testUser);
+        assertThrows(MindMyMoneyException.class, updateCommand::executeCommand);
+    }
+
+    /**
+     * Assert that when the update command fields are similar to the income in the list, an exception is thrown.
+     *
+     * */
```
Comment (khseah) [inline src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java:160]:
```suggestion
     **/
```

**Review by khseah — COMMENTED** [review]
Nice job in removing some of our bugs found during PED testing

File: src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java (hunk 5 above)
Comment (SeanHoWB) [reply src/test/java/seedu/mindmymoney/command/UpdateCommandTest.java]: wait. is it supposed to be just */ ?

File: src/main/java/seedu/mindmymoney/command/UpdateCommand.java (hunk 2 above)
Comment (limjierui) [reply src/main/java/seedu/mindmymoney/command/UpdateCommand.java]: resolved.

**Pull request merged**